dashboard/app/controllers/pd/mimeo_sso_controller.rb

```diff
@@ -0,0 +1,76 @@
+# Authenticate and redirect a Code Studio user to Mimeo using their single-sign-on (SSO) API
+class Pd::MimeoSsoController < ApplicationController
+  load_resource :enrollment, class: 'Pd::Enrollment', find_by: :code, id_param: :enrollment_code
+
+  ENDPOINT = 'https://my.mimeo.com/sso/authenticate.ashx'.freeze
+  AUTHORIZED_MARKETPLACE_URL = 'unpluggedlessonmaterials'.freeze
+  SSO_VERSION = '2.0'.freeze
+
+  ALLOWED_COURSES = [Pd::Workshop::COURSE_CSF].freeze
+
+  def authenticate_and_redirect
+    # Make sure the user is logged in and is the owner of this enrollment
+    authorize! :manage, @enrollment
+    raise CanCan::AccessDenied unless ALLOWED_COURSES.include? @enrollment.workshop.course
+
+    unless @enrollment.completed_survey?
+      # Until the survey is processed, it won't show up in the enrollment model.
+      # Check pegasus forms directly before giving up.
+      raise CanCan::AccessDenied unless PEGASUS_DB[:forms].where(kind: 'PdWorkshopSurvey', source_id: @enrollment.id).any?
+    end
+
+    # Set up fields for the Mimeo SSO API. Each of these will be rendered as a hidden field on a form which will be
+    # automatically posted to the @endpoint on page load.
+    # Mimeo's API will then authenticate and either redirect the user's browser to the redirectUrl on success,
+    # or to our errorUrl if it fails (see #error below).
+    @endpoint = ENDPOINT
+    @fields = {
+      firstName: encrypt_token(@enrollment.first_name),
+      lastName: encrypt_token(@enrollment.last_name),
+      email: encrypt_token(current_user.email),
```

Comment:
Are we guaranteed to have all these fields? It's possible for last_name to be blank if this is old (and arguably we shouldn't be requiring last name)